### PR TITLE
Better lstm

### PR DESCRIPTION
**Pull request by DaisukeBekki**
Torch.Tensor.Utilモジュールに unstack 関数を用意し、以下の関数を実装しました。それぞれ、入力が[Tensor]ではなく（orderがもう一つ高い）Tensorになっています。

LSTM.hs
- lstmLayers'

BiLSTM.ht
- biLstmLayer'
- biLstmLayers'

次のタスクは、上の（'付き）関数のテストを書くこと、および出力の型をpytorchと統一することです。

### CODE REVIEW

**Comment by bluesky0906** [comment]
LSTM.hsの方で、
lstmLayer'は層が重ねられないので、biLSTMと同様numOfLayersを指定できるようになると嬉しいです。

@DaisukeBekki

**Comment by DaisukeBekki** [comment]
まとめ：PyTorch-styleのLSTM（Torch.Layer.LSTM）
- bidirectionalかどうかをTrue/Falseで指定可
- num_layersをintで指定可（1未満を入力しても最低1層はある）
- dropoutProbをMaybe Doubleで指定可。Nothingならdropout layerなし。Just ...ならその値で各LSTM layerの入り口でdropout
- proj_sizeをMaybe Intで指定可。Nothingならprojection layerなし。Just...なら最後linear layerを加えて出力をその次元数×seq lengthに合わせる。
- c0, h0はLstmParamsには持たせず、外から(Tensor,Tensor)型の引数としてlstmLayer関数に与える。ここでLstmHypParamsと数字を合わせないといけない。そのサポートとして、InitialStatesHypParams型を作っておいた。
- 使い方のサンプルは、app/lstmSample/Main.hs を参照。

まだやるか迷っている点：
- bias: Boolを実装するかどうか。MaybeがParameterizedであることが分かったので技術的には可能。ただし、LinearLayerの仕様を変えることになる。
- batch処理と、batch_firstオプションを実装するかどうか。これは結構たいへん。

**Review by bluesky0906 — COMMENTED** [review]
ありがとうございます！
とりあえず、命名に関することだけコメントしました。

File: src/Torch/Layer/Linear.hs [hunk 1]
```diff
@@ -11,41 +11,52 @@ module Torch.Layer.Linear (
...
 import Torch.NN              (Parameter,Parameterized,Randomizable,sample)
 import Torch.Autograd        (IndependentTensor(..),makeIndependent)
-import Torch.Tensor.TensorFactories (randnIO')
-import Torch.Tensor.Initializers    (xavierUniform')
+import Torch.Tensor.TensorFactories (asTensor'',randintIO')
+--import Torch.Tensor.Initializers    (xavierUniform')
 
 data LinearHypParams = LinearHypParams {
   dev :: Device,
+  ifBias :: Bool,
```
[the remark this answers is not in the document]
Comment (DaisukeBekki) [reply src/Torch/Layer/Linear.hs:25]: hasBias、でどうでしょう？

File: src/Torch/Layer/LSTM.hs [hunk 2]
```diff
@@ -1,61 +1,172 @@
...
+          forwardLayer = fst $ unzip $ tail $ scanl' (lstmCell params) h0c0f inputs -- removing (h0,c0) by tail
+          backwardLayer = fst $ unzip $ init $ scanr (flip $ lstmCell params) h0c0b inputs -- removing (h0,c0) by init
+      return $ map (\(f,b)-> cat (Dim 0) [f,b]) $ zip forwardLayer backwardLayer
+    else do -- the case of LSTM
+      unless ((h0shape == [1,stateDim]) && (c0shape == [1,stateDim])) $ -- check the input shape
+        ioError $ userError $ "illegal LSTM shape of h0 or c0: " ++ (show h0shape) ++ " or " ++ (show c0shape)
+      let h0c0f = (select 0 0 h0, select 0 0 c0) 
+      return $ fst $ unzip $ tail $ scanl' (lstmCell params) h0c0f inputs -- | (c0,h0)は除くためtailを取る
+
+lstmLayers :: LstmHypParams -- ^ hyper parameters
```
Comment (bluesky0906) [inline src/Torch/Layer/LSTM.hs:127]: これはbidirectionalを使うために、hyper parameterを渡しているのでしょうか？
それならBoolだけ渡してしまう、配列の型からチェックする、または別名の関数を用意する方が良いと思います。

多くの場合、LSTMを用いたモデルを作るとき、
 - HypParams(Spec)にはLstmHypParamsそのものを持たないこと
 - model呼び出しする関数まで、HypParamsを渡し続けなければならないため不便です。

Comment (DaisukeBekki) [reply src/Torch/Layer/LSTM.hs]: lstmLayers (Torch.Layer.LSTMおよびTorch.Layer.SimpleLSTM）からLstmHypParams型の引数を除去しました。その代わり、lstmLayerがdropoutのための引数（Maybe Double型）を余計に取るように（そしてLstmHypParams型から同フィールドを削除するように）変更しました。使ってみてもらえたらと思います。

**Comment by bluesky0906** [comment]
lstmLayersがいくつか定義されてると思うので、その使い分けをコメントしてくださると助かります。

File: src/Torch/Layer/RNN.hs [hunk 3]
```diff
@@ -28,8 +28,8 @@ data RNNParams = RNNParams
 instance Randomizable RNNHypParams RNNParams where
```
Comment (bluesky0906) [inline src/Torch/Layer/RNN.hs:28]: RNNもLSTMと同様に
https://pytorch.org/docs/stable/generated/torch.nn.RNN.html
 - dropout
 - inputをTensorに
 - layerの数(今は必須ではない)
 - bidirectionalかどうか(今は必須ではない)
 を指定できると助かります。

Comment (bluesky0906) [reply src/Torch/Layer/RNN.hs:28]: @DaisukeBekki

**Comment by DaisukeBekki** [comment]
> lstmLayersがいくつか定義されてると思うので、その使い分けをコメントしてくださると助かります。

まず、Torch.Layer.ProtoType.LSTM は、古い実装を取っておいてあるだけなので、過去のバージョンが使いたいときのみ使ってください。

Torch.Layer.LSTMが現在のLSTMモジュールで、こちらのlstmLayers関数を使ってください。（SingleLstmLayerという関数がありますが、lstmLayersのサブルーチンであって、エクスポートはしていません）

Torch.Layer.SimpleLSTM にあるのは、h0,c0をLSTMのモデルに含んでしまうバージョンで、h0,c0を別途初期化するのが面倒な場合はこちらをどうぞ。Torch.Layer.LSTMの関数を呼び出して実装しているので内部の設計は同じです。

**Comment by bluesky0906** [comment]
> lstmLayersの使い分け

わからなくなりそうなので、コード内にコメント追加してほしいです:pray:

@DaisukeBekki

File: src/Torch/Layer/LSTM.hs [hunk 4]
```diff
@@ -1,61 +1,176 @@
...
+      unless ((h0shape == [1,stateDim]) && (c0shape == [1,stateDim])) $ -- check the input shape
+        ioError $ userError $ "illegal LSTM shape of h0 or c0: " ++ (show h0shape) ++ " or " ++ (show c0shape)
+      let h0c0f = (select 0 0 h0, select 0 0 c0) 
+      return $ fst $ unzip $ tail $ scanl' (lstmCell params) h0c0f inputs -- | (c0,h0)は除くためtailを取る
+
+lstmLayers :: LstmParams      -- ^ parameters (=model)
+  -> (Tensor,Tensor) -- ^ a pair of initial tensors: (D*numLayers,hiddenSize)
+  -> Maybe Double    -- ^ introduces a Dropout layer on the outputs of each LSTM layer except the last layer, with dropout probability equal to dropout.
+  -> Tensor          -- ^ an input tensor of shape (seqLen,inputSize)
+  -> Tensor          -- ^ [h_i] of shape (seqLen,D*hiddenSize)
```
[the remark this answers is not in the document]
Comment (DaisukeBekki) [reply src/Torch/Layer/LSTM.hs:132]: > lstmLayersがいくつか定義されてると思うので、その使い分けをコメントしてくださると助かります。

たしかにそうですね。メモを足しました。

File: src/Torch/Layer/SimpleLSTM.hs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+instance Randomizable LstmHypParams LstmParams where
+  sample LstmHypParams{..} = 
+    LstmParams
+      <$> (sample $ LSTM.LstmHypParams dev bidirectional inputSize hiddenSize numLayers hasBias projSize)
+      <*> (sample $ LSTM.InitialStatesHypParams dev bidirectional hiddenSize numLayers)
+
+lstmLayers :: LstmParams -- ^ params
+  -> Maybe Double -- ^ dropout
+  -> Tensor       -- ^ an input tensor of shape (L,H_in)
+  -> Tensor       -- ^ [his] of shape (L,D*H_out)
```
Comment (bluesky0906) [inline src/Torch/Layer/SimpleLSTM.hs:48]: こちらも同様、outputの変更をお願いします。

File: src/Torch/Layer/LSTM.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (DaisukeBekki) [reply src/Torch/Layer/LSTM.hs]: > ここcnも返してほしいです。 Pytorchと同じく(output, (h_n, c_n))にすると汎用性高そうです。

betterLSTMブランチにpushしました。

File: src/Torch/Layer/SimpleLSTM.hs (hunk 5 above)
Comment (DaisukeBekki) [reply src/Torch/Layer/SimpleLSTM.hs]: こちらも対応しています。

File: src/Torch/Layer/LSTM.hs [hunk 6]
```diff
@@ -1,61 +1,180 @@
...
+                       Nothing -> id
+      stackedLayers = \inputTensor -> foldl' 
+                                        (\hscs nextLayer -> (nextLayer . dropoutLayer . fst) hscs)
+                                        (firstLayer inputTensor)
+                                        restOfLayers
+      projLayer = case projParams of
+                    Just projP -> (stack (Dim 0)) . map (linearLayer projP) . unstack
+--                    Just projP -> linearLayer projP . reshape [] --broadcasting [oDim,projDim] to [seqLen,oDim]
+                    Nothing -> id
+  return $ (projLayer $ fst $ stackedLayers inputs, stackedLayers inputs)
```
Comment (bluesky0906) [inline src/Torch/Layer/LSTM.hs:158]: output、現在 ([seqLen,projSize],([seqLen,D*hiddenSize],[seqLen,D*hiddenSize]) 
となっていますが、pytorchに揃える(最終層の出力, (各層のh, 各層のc))の形にするのは難しいですか？

Comment (bluesky0906) [inline src/Torch/Layer/LSTM.hs:158]: `stackedLayers inputs` を2回書いていますがこれでは2回呼び出してしまわないでしょうか。

Comment (bluesky0906) [reply src/Torch/Layer/LSTM.hs:158]: @DaisukeBekki

Comment (DaisukeBekki) [reply src/Torch/Layer/LSTM.hs:158]: これはずさんでした。次のcommitで直します。

Comment (DaisukeBekki) [reply src/Torch/Layer/LSTM.hs]: > pytorchに揃える(最終層の出力, (各層・両方向のhn(最終状態), 各層・両方向ののcn))の形にする

この形に修正しました。この修正についてのテスト、およびコメントの訂正がまだ不十分ですが、順次対応します。

**Comment by DaisukeBekki** [comment]
lstmLayerがshapeについての全テストに通るようになりました。テストは

/test/singleLstmLayerTest/Main.hs
/test/lstmLayerTest/Main.hs

にあり、lstmLayer引数の全組み合わせについて、出力shapeのテストと微分可能性のテストを兼ねています。現バージョンはこれらに通っているので、ある程度頑健になったと思いますが、もし使っていてエラーが出たらご報告ください。

また、現バージョンでは再帰の際にむやみにリスト化せず、できるだけTensorのまま処理するようにしたので、少しは速くなったのではないかと期待しています。

**Comment by DaisukeBekki** [comment]
master branchにmergeしました。
